src/api-binder.ts

```diff
@@ -0,0 +1,890 @@
+				await this.config.regenerateRegistrationFields();
+			}
+			throw e;
+		}
+	}
+
+	private async provision() {
+		let device: Device | null = null;
+		// FIXME: Config typing
+		const opts = ((await this.config.get(
+			'provisioningOptions',
+		)) as any) as Dictionary<any>;
+
+		if (
+			opts.registered_at != null &&
+			opts.deviceId != null &&
+			opts.provisioningApiKey == null
+		) {
+			return;
+		}
+
+		if (opts.registered_at != null && opts.deviceId == null) {
+			console.log(
+				'Device is registered but no device id available, attempting key exchange',
+			);
+			device =
+				(await this.exchangeKeyAndGetDeviceOrRegenerate(
+					opts as KeyExchangeOpts,
+				)) || null;
+		} else if (opts.registered_at == null) {
+			console.log('New device detected. Provisioning...');
+			try {
+				device = await deviceRegister.register(opts).timeout(opts.apiTimeout);
+				opts.registered_at = Date.now();
+			} catch (err) {
+				if (DuplicateUuidError(err)) {
+					console.log('UUID already registered, trying a key exchange');
+					await this.exchangeKeyAndGetDeviceOrRegenerate(
+						opts as KeyExchangeOpts,
+					);
+				} else {
+					throw err;
+				}
+			}
+		} else if (opts.provisioningApiKey != null) {
+			console.log(
+				'Device is registered but we still have an apiKey, attempting key exchange',
+			);
+			device = await this.exchangeKeyAndGetDevice(opts as KeyExchangeOpts);
+		}
+
+		if (!device) {
+			// TODO: Type this?
+			throw new Error(`Failed to provision device!`);
+		}
+		const { id } = device;
+		if (!this.balenaApi) {
+			throw new InternalInconsistencyError(
+				'Attempting to provision a device without an initialized API client',
+			);
+		}
+		this.balenaApi.passthrough.headers.Authorization = `Bearer ${
+			opts.deviceApiKey
+		}`;
+
+		const configToUpdate = {
+			registered_at: opts.registered_at,
+			deviceId: id,
+			apiKey: null,
+		};
+		await this.config.set(configToUpdate);
+		this.eventTracker.track('Device bootstrap success');
+
+		// Now check if we need to pin the device
+		const toPin = await this.config.get('pinDevice');
+		let pinValue: DevicePinInfo | null = null;
+		try {
+			pinValue = JSON.parse(toPin as string);
+		} catch (e) {
+			console.log('Warning: Malformed pinDevice value in supervisor database');
+		}
+
+		if (pinValue != null) {
+			if (pinValue.app == null || pinValue.commit == null) {
+				console.log(
+					`Malformed pinDEvice fields in supervisor database: ${pinValue}`,
+				);
+				return;
+			}
+			console.log('Attempting to pin device to preloaded release...');
+			return this.pinDevice(pinValue);
+		}
+	}
+
+	private async provisionOrRetry(retryDelay: number): Promise<void> {
+		this.eventTracker.track('Device bootstrap');
+		try {
+			await this.provision();
+		} catch (e) {
+			this.eventTracker.track(`Device bootstrap failed, retrying`, {
+				error: e,
+				delay: retryDelay,
+			});
+			return Bluebird.delay(retryDelay).then(() => {
```

Comment:
This can be
```suggestion
			await Bluebird.delay(retryDelay);
```
to keep the style more consistent